test/CPUActivityAgentTest.cpp

```diff
@@ -0,0 +1,583 @@
+
+    m_cpu_freq_min = 1000000000.0;
+    m_cpu_freq_max = 3700000000.0;
+    m_cpu_uncore_freq_min = 1200000000.0;
+    m_cpu_uncore_freq_max = 2400000000.0;
+
+    ON_CALL(*m_platform_io, control_domain_type("CPU_FREQUENCY_MAX_CONTROL"))
+            .WillByDefault(Return(GEOPM_DOMAIN_CPU));
+    ON_CALL(*m_platform_io, signal_domain_type("MSR::CPU_SCALABILITY_RATIO"))
+            .WillByDefault(Return(GEOPM_DOMAIN_CPU));
+
+    ON_CALL(*m_platform_io, read_signal("CPU_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0))
+            .WillByDefault(Return(m_cpu_freq_min));
+    ON_CALL(*m_platform_io, read_signal("CPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0))
+            .WillByDefault(Return(m_cpu_freq_max));
+
+    ON_CALL(*m_platform_io, read_signal("CPU_UNCORE_FREQUENCY_MIN_CONTROL", GEOPM_DOMAIN_BOARD, 0))
+            .WillByDefault(Return(m_cpu_uncore_freq_min));
+    ON_CALL(*m_platform_io, read_signal("CPU_UNCORE_FREQUENCY_MAX_CONTROL", GEOPM_DOMAIN_BOARD, 0))
+            .WillByDefault(Return(m_cpu_uncore_freq_max));
+
+    ASSERT_LT(m_cpu_freq_min, 2e9);
+    ASSERT_LT(m_cpu_freq_max, 4e9);
+    ASSERT_LT(m_cpu_freq_min, m_cpu_freq_max);
+    ASSERT_LT(m_cpu_uncore_freq_min, 2e9);
+    ASSERT_LT(m_cpu_uncore_freq_max, 3e9);
+    ASSERT_LT(m_cpu_uncore_freq_min, m_cpu_uncore_freq_max);
+
+    EXPECT_CALL(*m_platform_io, write_control("MSR::PQR_ASSOC:RMID", _, _, _)).Times(1);
+    EXPECT_CALL(*m_platform_io, write_control("MSR::QM_EVTSEL:RMID", _, _, _)).Times(1);
+    EXPECT_CALL(*m_platform_io, write_control("MSR::QM_EVTSEL:EVENT_ID", _, _, _)).Times(1);
+    EXPECT_CALL(*m_platform_io, read_signal("CPU_UNCORE_FREQUENCY_MIN_CONTROL", _, _)).Times(1);
+    EXPECT_CALL(*m_platform_io, read_signal("CPU_UNCORE_FREQUENCY_MAX_CONTROL", _, _)).Times(1);
+    EXPECT_CALL(*m_platform_io, read_signal("CPU_FREQUENCY_MIN_AVAIL", _, _)).Times(1);
+    EXPECT_CALL(*m_platform_io, read_signal("CPU_FREQUENCY_MAX_AVAIL", _, _)).Times(1);
+
+    m_agent = geopm::make_unique<CPUActivityAgent>(*m_platform_io, *m_platform_topo);
+    m_num_policy = m_agent->policy_names().size();
+
+    m_default_policy = {m_cpu_freq_max, m_cpu_freq_min, m_cpu_uncore_freq_max,
+                        m_cpu_uncore_freq_min, NAN};
+
+    m_cpu_uncore_freqs = {1.2e9, 1.3e9, 1.4e9, 1.5e9, 1.6e9, 1.7e9, 1.8e9,
+                      1.9e9, 2.0e9, 2.1e9, 2.2e9, 2.3e9, 2.4e9};
+    m_mbm_max = {45414967307.69231, 64326515384.61539, 72956528846.15384,
+                 77349315384.61539, 82345998076.92308, 87738286538.46153,
+                 91966364814.81482, 96728174074.07408, 100648379629.62962,
+                 102409246296.2963, 103624103703.7037, 104268944444.44444,
+                 104748888888.88889};
+    ASSERT_EQ(m_cpu_uncore_freqs.size(), m_mbm_max.size());
+    ASSERT_EQ(m_mbm_max.size(), M_NUM_UNCORE_MBM_READINGS);
+
+    for (size_t i = 0; i < M_NUM_UNCORE_MBM_READINGS; ++i) {
+        m_default_policy.push_back(static_cast<double>(m_cpu_uncore_freqs[i]));
```

Comment:
Why is the static_cast necessary? m_cpu_uncore_freqs is already a vector of doubles.